### PR TITLE
matching: use polymorphic variants to classify general/half_simple/simple patterns

### PR DESCRIPTION
**Pull request by trefis**
Another one-commit-PR extracted from #9321 .

Note: we have tried an alternative to this, which was to make the "intermediary types" (i.e. the ones representing half-simplified and simplified patterns) abstract.
But it didn't turn out as nice as this.

### CODE REVIEW

File: lambda/matching.ml [hunk 1]
```diff
@@ -210,73 +269,80 @@ end = struct
...
+          `Construct (cstr, cst_descr, List.map (alpha_pat env) args)
+      | `Variant (cstr, argo, row_desc) ->
+          `Variant (cstr, Option.map (alpha_pat env) argo, row_desc)
+      | `Record (fields, closed) ->
+          let alpha_field env (lid, l, p) = (lid, l, alpha_pat env p) in
+          `Record (List.map (alpha_field env) fields, closed)
+      | `Array ps -> `Array (List.map (alpha_pat env) ps)
+      | `Lazy p -> `Lazy (alpha_pat env p)
+    in
+    { p with pat_desc }
```
Comment (Octachron) [inline lambda/matching.ml:346]: This feels like the inlined version of a `fmap: (Typedtree.pattern -> Typedtree.pattern) -> pattern -> pattern` function. I would prefer to have the generic function explictily defined.

File: lambda/matching.ml [hunk 2]
```diff
@@ -1132,10 +1199,10 @@ module Or_matrix = struct
...
     (not (is_guarded act)) && Parmatch.le_pats ps qs
 
-  let safe_below_or_matrix to_pattern l (q, qs) =
+  let safe_below_or_matrix l (q, qs) =
     List.for_all
       (fun ((p, ps), act_p) ->
-        let p = to_pattern p in
+        let p = General.erase p in
         match p.pat_desc with
         | Tpat_or _ -> disjoint p q || safe_below (ps, act_p) qs
```
Comment (Octachron) [inline lambda/matching.ml:1208]: Erasing `p` just before pattern matching on it seems wasteful? Could we commute the erasing and the pattern matching?

File: lambda/matching.ml (hunk 1 above)
Comment (trefis) [reply lambda/matching.ml:346]: Done.

File: lambda/matching.ml (hunk 2 above)
Comment (trefis) [reply lambda/matching.ml:1208]: Done.

**Comment by trefis** [comment]
I discussed this offline with @Octachron last week, and he does not seem particularly excited by the use of polymorphic variants.
He's implemented [an alternative](https://github.com/Octachron/ocaml/commit/02589e16d5a50a0ca200e7afeda3fa3b6d8f097c) to this PR, which seems to handle our needs.

Having just read over both versions of `matching.ml` (Octachron's one, and the one in this PR) I have a slight preference for the version with polymorphic variant; I might of course be biased, but I do not think so:
- I was initially reluctant myself when @gasche proposed it
- polymorphic variants are perfectly adapted to the problem at hand (more so than phantom encodings)
- reading `precompile_or` written using polymorphic variant gives me a weirdly satisfying feeling

That said, I will let @Octachron present his grievances, and defend the merits of his approach.

**Comment by Octachron** [comment]
My main grievance is that the polymorphic variant types are not used that much in this PR in fine.

My view of this PR is that it refines the pattern type family in order to both expose the subtyping relationship between the family members and precise their "decomposition" (for lack of better or).

The first part can be done without polymorphic variants (as seen in my toy implementation), so the real advantages of polymorphic variants comes from the second part.

However, most of the PR improvement comes from exposing the first part: the subtyping relationship makes it possible to have a generic `General.erase` that works on all type in the family.

The precise decomposition of polymorphic variant is used exactly thrice in the PR. Two of those are an effect of making `try_not_or` unnecessary which is certainly nice. Unfortunately, everywhere else, we end up with a pattern of first refining a pattern, doing a small amount of works on the refined pattern and then we are blocked because the next step cannot handle the refined types so we need to undo our refining work.

I ended up looking to the overview pattern matching PR to see if it is only a short growing pain. However, even there, it is not that clear: the pattern of refine/erase is still present.

In other words, I am paradoxically stating that the lack of ambition in this PR makes it harder to review. And I am not fully convinced that this back-and-forth is only a growing pain, even if polymorphic variants does seem to be a natural fit.

Overall, I think that the balance still tip in favor of this PR, but by a far thinner margin than I expected when I started the review.

**Comment by gasche** [comment]
I don't like @Octachron's proposal for two reasons:

1. It feels more complex. We are encoding a subtyping relation using private synonyms within phantom type parameters. We know that this corner of the type system can express rich subtyping relations, but it does not mean that the result is pretty. I personally find `simple half t` or `non_simple half t` to be on the confusing side.

2. Most importantly, it provides less static guarantees. As with all solutions based on phantom types, the safety comes from the module type boundary: code outside the boundary is safe *provided* that the interface invariants are properly satisfied, and code inside the boundary has to be trusted, it is never checked for safety. On the contrary, our polymorphic-variant approach performs a precise static analysis in the language directly, instead of being encoded in an axiomatized form in the module interface. (This is the same reason why GADTs are typically safer and better than phantom types.)

Compare for example:

```ocaml
val explode : Half_simple.pattern -> Ident.t list -> Simple.clause list -> Simple.clause list

(* polymorphic variant version -- slightly reworded from the PR *)
let rec explode p aliases rem =
  let explode_p p aliases rem = explode (General.view p) aliases rem in
  match p.pat_desc with
  | `Or (p1, p2, _) -> explode_p p1 aliases (explode_p p2 aliases rem)
  | `Alias (p, id, _) -> explode_p p (id :: aliases) rem
  | `Var (id, str) ->
      let env = mk_alpha_env arg (x :: aliases) vars in
      ((omega, patl), mk_action ~vars:(List.map snd env)) :: rem
  | #simple_view as view ->
      let p = { p with pat_desc = view } in
      let env = mk_alpha_env arg aliases vars in
      ((alpha env p, patl), mk_action ~vars:(List.map snd env)) :: rem

(* phantom type versions *)
let rec explode p aliases rem =
   match p.pat_desc with
   | Tpat_or (p1, p2, _) -> explode p1 aliases (explode p2 aliases rem)
   | Tpat_alias (p, id, _) -> explode p (id :: aliases) rem
   | Tpat_var (x, _) ->
       let env = mk_alpha_env arg (x :: aliases) vars in
       ((omega, patl), mk_action ~vars:(List.map snd env)) :: rem
   | _ ->
       let env = mk_alpha_env arg aliases vars in
       ((alpha_pat env p, patl), mk_action ~vars:(List.map snd env)) :: rem
```

In the polymorphic version, we know statically that the `#simple_view as view` matches only simple pattern descriptions, and we retype `p` with this finer-grained type information. If we had forgotten one of the non-simple cases above, we would get a type error. With the phantom type version, we have to trust this code (which sits within the abstraction boundary), and forgetting a case would result in a silent breach of static enforcement.

There are more such code traversals in matching.ml and parmatch.ml, that can benefit from the static safety of the polymorphic-variant version. Most of these functions should not be pushed within the abstraction boundary. One could define a generic "non-simple decomposition" higher-order function that would sit inside the abstraction, to be used by those traversals outside, but then again this would result in more complex code.

> I ended up looking to the overview pattern matching PR to see if it is only a short growing pain. However, even there, it is not that clear: the pattern of refine/erase is still present.

Sure, there is a frontier between the static-discipline-aware code (that knows which patterns are (half)simple in its inputs and outputs) and the unaware code. Whenever we call an unaware helper function (for example `Typedtree.alpha_pat`), there is some unpleasant ceremony to cross the frontier (typically an erasing coercion on the way out, and an unsafe cast on the way back out; or duplication of the code within the frontier). Your approach slightly reduces this ceremony because fine-grained patterns are subtypes of ordinary patterns, but it does not remove it. We could also reduce it by pushing more pattern-processing code to be discipline-aware, or by changing the Typedtree pattern definition to be a polymorphic variant from the start (so that we regain the erasure-is-subtyping convenience). I am not in favor of the latter (I would rather keep the entry cost of polymorphic variants only in the places where they are really useful), and I think that the former will come over time. Still during the transition period (which is, indeed, not finished by the end of our current PRs) I prefer our approach to yours, because it provides more type-safety.

File: lambda/matching.ml [hunk 3]
```diff
@@ -210,73 +269,82 @@ end = struct
...
+      in
+      match p.pat_desc with
+      | `Any -> stop p `Any
+      | `Var (id, s) -> continue p (`Alias (omega, id, s))
+      | `Alias (p, id, _) ->
+          let k = Typeopt.value_kind p.pat_env p.pat_type in
+          aux
+            ( (General.view p, patl),
+              bind_with_value_kind Alias (id, k) arg action )
+      | `Record ([], _) as view -> stop p view
```
Comment (Octachron) [inline lambda/matching.ml:288]: The head of this pattern changed from `` `Any `` to `` `Record ``. Since this case cannot happen in the surface language, this changes is undetectable. However, I would prefer to keep the previous behavior to make sure that all records are extended in half compiler pattern are fully extended (it looks like extending them is not an option since `all_record_args` requires on label to compute the list of all labels).

Comment (gasche) [reply lambda/matching.ml:288]: Maybe we could try to expand record labels from the `pat_type` rather than a label? (Using `Any` here has the property that you give, but it is also wrong semantically, at least in an untyped model.)

File: lambda/matching.ml [hunk 4]
```diff
@@ -290,25 +358,25 @@ end = struct
...
-          let env = mk_alpha_env arg (x :: aliases) vars in
-          ((omega, patl), mk_action ~vars:(List.map snd env)) :: rem
-      | _ ->
+      | `Or (p1, p2, _) ->
+          split_explode p1 aliases (split_explode p2 aliases rem)
+      | `Alias (p, id, _) -> split_explode p (id :: aliases) rem
+      | `Var (id, str) ->
+          explode
+            { p with pat_desc = `Alias (Parmatch.omega, id, str) }
+            aliases rem
```
Comment (Octachron) [inline lambda/matching.ml:371]: The behavior is the same as before this PR, but the previous version inlined the recursive call to `explode` whereas the new code goes through an `` `Alias `` intermediary stage. I imagine that this is done to follow the code path in `Half_simplified.of_clause`?

File: lambda/matching.ml [hunk 5]
```diff
@@ -210,73 +269,82 @@ end = struct
...
     in
     aux cl
 end
 
 exception Cannot_flatten
 
 module Simple : sig
-  type pattern
-  (** A fully simplified pattern: or-patterns have been exploded, and the
-      remaining aliases have been removed and replaced by bindings in actions *)
```
Comment (Octachron) [inline lambda/matching.ml:251]: The documentation comment is certainly less necessary, but it seems still useful to document how we removed or-patterns and aliases?

File: lambda/matching.ml [hunk 6]
```diff
@@ -142,19 +142,82 @@ let all_record_args lbls =
...
+    | `Variant (cstr, arg, row_desc) -> Tpat_variant (cstr, arg, row_desc)
+    | `Record (fields, closed) -> Tpat_record (fields, closed)
+    | `Array ps -> Tpat_array ps
+    | `Or (p, q, row_desc) -> Tpat_or (p, q, row_desc)
+    | `Lazy p -> Tpat_lazy p
+
+  let erase p = { p with pat_desc = erase_desc p.pat_desc }
 end
 
+let omega_ : [> `Any ] pattern_data = { Parmatch.omega with pat_desc = `Any }
```
Comment (Octachron) [inline lambda/matching.ml:219]: It may be nice to have a better qualifier than `_` for this `omega`. `General.omega`? `omega_view`? `just_omega`? `exactly_omega`?

File: lambda/matching.ml (hunk 3 above)
Comment (Octachron) [reply lambda/matching.ml:288]: I just remarked that the `simplify_under_or` function raises a fatal_error on empty record patterns. So it definitely does not make sense to spend too much type decorating this specific empty set.

File: lambda/matching.ml (hunk 4 above)
Comment (gasche) [reply lambda/matching.ml:371]: The idea was that it was clearer to just rewrite `x` into `_ as x` on the fly than to copy the implementation of the Alias and omega cases in the Var case.

File: lambda/matching.ml (hunk 5 above)
Comment (gasche) [reply lambda/matching.ml:251]: This is not about the structure of simple patterns; if we want to document this, I guess it would go in the half-simplification function for clauses.

**Review by Octachron — APPROVED** [review]
It seems indeed more natural to let the type-level computations to the type checker rather than to encode them as admitted axioms in the module interface.

My remaining remarks are completely minor.

Like I said, the subtyping and the removal of `try_no_or` are certainly nice, thus the PR seems good to go for me.

File: lambda/matching.ml (hunk 5 above)
Comment (gasche) [reply lambda/matching.ml:251]: I am turning this into a comment on `Half_simple.of_clause`.

File: lambda/matching.ml (hunk 6 above)
Comment (gasche) [reply lambda/matching.ml:219]: The names you give are not clearly better (`General.omega` is probably the best candidate). This declaration is gone in the final PR state of #9321, replaced by `Simple.omega : [> Simple.view ] pattern`, so I would propose to do nothing for now.

Comment (Octachron) [reply lambda/matching.ml:219]: Indeed, `omega_` is a fine ephemeral name.

**Comment by gasche** [comment]
@trefis I rebased this PR on top of trunk (most of the work was coming from changes to previous PRs in the series). It's now up to you to decide when to merge (assuming CI is green), and to shoot the next round.

**Comment by trefis** [comment]
Thanks!

I'll probably send the next PR early next week.
